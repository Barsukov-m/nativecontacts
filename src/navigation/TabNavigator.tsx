import { createBottomTabNavigator } from '@react-navigation/bottom-tabs';
import ContactsStack from './ContactsStack';
import NewContactStack from './NewContactStack';
import MeStack from './MeStack';
import Icon from 'react-native-vector-icons/MaterialIcons';
import themeColors from '../constants/themeColors';

const Tab = createBottomTabNavigator();

const TabNavigator: React.FC = () => {
	const tabOptions = {
		headerShown: false,
		tabBarActiveTintColor: themeColors.tabTextActive,
		tabBarInactiveTintColor: themeColors.tabTextInactive,
		tabBarActiveBackgroundColor: themeColors.tabBackgroundActive,
		tabBarInactiveBackgroundColor: themeColors.tabBackgroundInactive,
		tabBarLabelStyle: {
			fontSize: 10,
			paddingBottom: 5,
		},
	};

	return (
		<Tab.Navigator initialRouteName="Contacts" screenOptions={() => tabOptions}>
			<Tab.Screen
				name="ContactsTab"
				component={ContactsStack}
				options={{
					tabBarLabel: 'Contacts',
					tabBarIcon: ({ color }) => (
						<Icon name="contacts" color={color} size={20} />
					),
				}}
			/>
			<Tab.Screen
				name="NewContactTab"
				component={NewContactStack}
				options={{
					tabBarLabel: 'New Contact',
					tabBarIcon: ({ color }) => (
						<Icon name="person-add" color={color} size={20} />
					),
				}}
			/>
			<Tab.Screen
				name="MeTab"
				component={MeStack}
				options={{
					tabBarLabel: 'Me',
					tabBarIcon: ({ color }) => (
						<Icon name="person" color={color} size={20} />
					),
				}}
			/>
		</Tab.Navigator>
	);
};

export default TabNavigator;
