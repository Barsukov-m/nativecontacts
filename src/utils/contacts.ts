import { Alert } from 'react-native';
import { ContactInterface } from '../types/navigationTypes';

export const capitalizeFirstLetter = (text: string): string => {
  return text[0].toUpperCase() + text.slice(1);
};

export const getHeaderTitleByContact = (contact: ContactInterface) => {
  const firstName = capitalizeFirstLetter(contact.name.first);
  const lastName = capitalizeFirstLetter(contact.name.last);
  return `${firstName} ${lastName}`;
};

export const wipWarning = () => {
  Alert.alert('Work in Progress ğŸ¤«');
};
