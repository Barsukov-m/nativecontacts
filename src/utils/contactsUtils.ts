import { Alert } from 'react-native';
import { ContactInterface } from '../types/navigationTypes';

export const capitalizeFirstLetter = (text: string): string => {
	return text[0].toUpperCase() + text.slice(1);
};

export const getHeaderTitleByContact = (contact: ContactInterface) => {
	const firstName = contact.firstName
		? capitalizeFirstLetter(contact.firstName)
		: '';
	const lastName = contact.lastName
		? capitalizeFirstLetter(contact.lastName)
		: '';
	return `${firstName} ${lastName}`;
};

export const logWarning = (title: string, message: string) => {
	Alert.alert(title, message);
};

export const wipWarning = () => {
	logWarning('Work in Progress ğŸ¤«', 'This feature is currently in development');
};
