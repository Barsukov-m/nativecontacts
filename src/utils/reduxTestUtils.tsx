import { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
import { render, renderHook, waitFor } from '@testing-library/react-native';
import type { RenderOptions } from '@testing-library/react-native';
import type { PreloadedState } from '@reduxjs/toolkit';
import { setupStore, type AppStore, type RootState } from '../store/index';
import { useFetchContactsQuery } from 'store/apis/contacts';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
	preloadedState?: PreloadedState<RootState>;
	store?: AppStore;
}

export const renderWithProviders = (
	ui: React.ReactElement,
	{
		preloadedState = {} as PreloadedState<RootState>,
		store = setupStore(preloadedState),
		...renderOptions
	}: ExtendedRenderOptions = {}
) => {
	const Wrapper = ({
		children,
	}: PropsWithChildren<{ store: AppStore }>): JSX.Element => {
		return <Provider store={store}>{children}</Provider>;
	};

	return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};

export const renderHookWithProviders = <Response, Params>(
	hook: (args: Params) => Response,
	{
		preloadedState,
		store = setupStore(preloadedState),
		...renderOptions
	}: ExtendedRenderOptions = {}
) => {
	const Wrapper = ({
		children,
	}: PropsWithChildren<{ store: AppStore }>): JSX.Element => {
		return <Provider store={store}>{children}</Provider>;
	};
	return renderHook(hook, {
		wrapper: Wrapper,
		...renderOptions,
	});
};

export const renderContactsOnScreen = async () => {
	const { result } = renderHookWithProviders(() => useFetchContactsQuery());
	await waitFor(() => expect(result.current.isFetching).toBe(false));
};
