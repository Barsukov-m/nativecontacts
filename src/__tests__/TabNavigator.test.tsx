import { NavigationContainer } from '@react-navigation/native';
import Icon from 'react-native-vector-icons/MaterialIcons';
import TabNavigator from '../navigation/TabNavigator';
import { renderWithProviders } from 'utils/reduxTestUtils';

jest.mock('react-native-vector-icons/MaterialIcons');

describe('<TabNavigator />', () => {
	const renderComponent = () =>
		renderWithProviders(
			<NavigationContainer>
				<TabNavigator />
			</NavigationContainer>
		);

	it('renders correctly', () => {
		const tree = renderComponent();

		expect(tree).toMatchSnapshot();
	});

	it('renders the Contacts tab with correct icon and label', () => {
		const { getByText } = renderComponent();

		// Check for the label
		expect(getByText('Contacts')).toBeTruthy();

		// Check for the icon name
		expect(Icon).toHaveBeenCalledWith(
			expect.objectContaining({
				name: 'contacts',
			}),
			{}
		);
	});

	it('renders the New Contact tab with correct icon and label', () => {
		const { getByText } = renderComponent();

		expect(getByText('New Contact')).toBeTruthy();

		expect(Icon).toHaveBeenCalledWith(
			expect.objectContaining({
				name: 'person-add',
			}),
			{}
		);
	});

	it('renders the Me tab with correct icon and label', () => {
		const { getByText } = renderComponent();

		expect(getByText('Me')).toBeTruthy();

		expect(Icon).toHaveBeenCalledWith(
			expect.objectContaining({
				name: 'person',
			}),
			{}
		);
	});
});
