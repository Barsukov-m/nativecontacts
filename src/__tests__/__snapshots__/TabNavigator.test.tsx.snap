// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TabNavigator /> renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    {
      "flex": 1,
    }
  }
>
  <RNSScreenNavigationContainer
    style={
      {
        "flex": 1,
        "overflow": "hidden",
      }
    }
  >
    <RNSScreen
      activityState={2}
      collapsable={false}
      gestureResponseDistance={
        {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      onGestureCancel={[Function]}
      sheetAllowedDetents="large"
      sheetCornerRadius={-1}
      sheetExpandsWhenScrolledToEdge={true}
      sheetGrabberVisible={false}
      sheetLargestUndimmedDetent="all"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "zIndex": 0,
        }
      }
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          [
            {
              "backgroundColor": "rgb(242, 242, 242)",
              "flex": 1,
            },
            [
              {
                "flex": 1,
                "flexDirection": "column-reverse",
              },
              undefined,
            ],
          ]
        }
      >
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <View
            style={
              [
                {
                  "flex": 1,
                },
                undefined,
              ]
            }
          >
            <RNSScreenStack
              style={
                {
                  "flex": 1,
                }
              }
            >
              <RNSScreen
                collapsable={false}
                gestureResponseDistance={
                  {
                    "bottom": -1,
                    "end": -1,
                    "start": -1,
                    "top": -1,
                  }
                }
                nativeBackButtonDismissalEnabled={false}
                onAppear={[Function]}
                onDisappear={[Function]}
                onDismissed={[Function]}
                onGestureCancel={[Function]}
                onHeaderBackButtonClicked={[Function]}
                onNativeDismissCancelled={[Function]}
                onTransitionProgress={[Function]}
                onWillDisappear={[Function]}
                replaceAnimation="push"
                sheetAllowedDetents="large"
                sheetCornerRadius={-1}
                sheetExpandsWhenScrolledToEdge={true}
                sheetGrabberVisible={false}
                sheetLargestUndimmedDetent="all"
                stackPresentation="push"
                style={
                  {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
                swipeDirection="horizontal"
              >
                <RNSScreenStackHeaderConfig
                  backButtonInCustomView={false}
                  backTitleVisible={true}
                  backgroundColor="rgb(255, 255, 255)"
                  color="rgb(0, 122, 255)"
                  direction="ltr"
                  disableBackButtonMenu={false}
                  hidden={true}
                  hideBackButton={false}
                  largeTitleHideShadow={false}
                  title="Home"
                  titleColor="rgb(28, 28, 30)"
                  topInsetEnabled={false}
                  translucent={false}
                />
                <View
                  accessibilityElementsHidden={false}
                  importantForAccessibility="auto"
                  style={
                    {
                      "flex": 1,
                      "flexDirection": "column-reverse",
                    }
                  }
                >
                  <View
                    style={
                      [
                        {
                          "flex": 1,
                        },
                        {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ]
                    }
                  >
                    <View
                      style={
                        {
                          "alignItems": "center",
                          "height": "100%",
                          "justifyContent": "center",
                        }
                      }
                    >
                      <ActivityIndicator
                        size="large"
                      />
                    </View>
                  </View>
                  <View
                    onLayout={[Function]}
                    style={null}
                  >
                    <View
                      style={
                        {
                          "alignItems": "center",
                          "backgroundColor": "#FAFAFA",
                          "elevation": 5,
                          "flexDirection": "row",
                          "height": 56,
                          "justifyContent": "space-between",
                          "paddingHorizontal": 16,
                          "shadowColor": "#212121",
                          "shadowOffset": {
                            "height": 2,
                            "width": 0,
                          },
                          "shadowOpacity": 0.25,
                          "shadowRadius": 3.84,
                        }
                      }
                    >
                      <View
                        style={
                          {
                            "flexDirection": "row",
                          }
                        }
                      >
                        <Text>
                          Icon
                        </Text>
                        <Text
                          style={
                            {
                              "color": "#212121",
                              "fontSize": 20,
                              "fontWeight": "500",
                            }
                          }
                        >
                          Native Contacts
                        </Text>
                      </View>
                    </View>
                  </View>
                </View>
              </RNSScreen>
            </RNSScreenStack>
          </View>
        </View>
      </View>
    </RNSScreen>
  </RNSScreenNavigationContainer>
  <View
    collapsable={false}
    onLayout={[Function]}
    pointerEvents="auto"
    style={
      {
        "backgroundColor": "rgb(255, 255, 255)",
        "borderTopColor": "rgb(216, 216, 216)",
        "borderTopWidth": 0.5,
        "bottom": 0,
        "elevation": 8,
        "height": 49,
        "left": 0,
        "paddingBottom": 0,
        "paddingHorizontal": 0,
        "position": null,
        "right": 0,
        "transform": [
          {
            "translateY": 0,
          },
        ],
      }
    }
  >
    <View
      pointerEvents="none"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
    <View
      accessibilityRole="tablist"
      style={
        {
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        accessibilityLabel="Contacts, tab, 1 of 3"
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": true,
          }
        }
        accessibilityStates={
          [
            "selected",
          ]
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "flex": 1,
            },
            {
              "backgroundColor": "#1976D2",
            },
            {
              "flexDirection": "column",
              "justifyContent": "flex-end",
            },
            undefined,
          ]
        }
      >
        <View
          style={
            [
              {
                "flex": 1,
              },
              undefined,
            ]
          }
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "alignSelf": "center",
                  "height": "100%",
                  "justifyContent": "center",
                  "minWidth": 25,
                  "position": "absolute",
                  "width": "100%",
                },
                {
                  "opacity": 1,
                },
              ]
            }
          >
            <Text>
              Icon
            </Text>
          </View>
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "alignSelf": "center",
                  "height": "100%",
                  "justifyContent": "center",
                  "minWidth": 25,
                  "position": "absolute",
                  "width": "100%",
                },
                {
                  "opacity": 0,
                },
              ]
            }
          >
            <Text>
              Icon
            </Text>
          </View>
        </View>
        <Text
          numberOfLines={1}
          style={
            [
              {
                "backgroundColor": "transparent",
                "textAlign": "center",
              },
              {
                "color": "#FAFAFA",
              },
              {
                "fontSize": 10,
              },
              {
                "fontSize": 10,
                "paddingBottom": 5,
              },
            ]
          }
        >
          Contacts
        </Text>
      </View>
      <View
        accessibilityLabel="New Contact, tab, 2 of 3"
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": false,
          }
        }
        accessibilityStates={[]}
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "flex": 1,
            },
            {
              "backgroundColor": "#2196F3",
            },
            {
              "flexDirection": "column",
              "justifyContent": "flex-end",
            },
            undefined,
          ]
        }
      >
        <View
          style={
            [
              {
                "flex": 1,
              },
              undefined,
            ]
          }
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "alignSelf": "center",
                  "height": "100%",
                  "justifyContent": "center",
                  "minWidth": 25,
                  "position": "absolute",
                  "width": "100%",
                },
                {
                  "opacity": 0,
                },
              ]
            }
          >
            <Text>
              Icon
            </Text>
          </View>
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "alignSelf": "center",
                  "height": "100%",
                  "justifyContent": "center",
                  "minWidth": 25,
                  "position": "absolute",
                  "width": "100%",
                },
                {
                  "opacity": 1,
                },
              ]
            }
          >
            <Text>
              Icon
            </Text>
          </View>
        </View>
        <Text
          numberOfLines={1}
          style={
            [
              {
                "backgroundColor": "transparent",
                "textAlign": "center",
              },
              {
                "color": "#BBDEFB",
              },
              {
                "fontSize": 10,
              },
              {
                "fontSize": 10,
                "paddingBottom": 5,
              },
            ]
          }
        >
          New Contact
        </Text>
      </View>
      <View
        accessibilityLabel="Me, tab, 3 of 3"
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": false,
          }
        }
        accessibilityStates={[]}
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "flex": 1,
            },
            {
              "backgroundColor": "#2196F3",
            },
            {
              "flexDirection": "column",
              "justifyContent": "flex-end",
            },
            undefined,
          ]
        }
      >
        <View
          style={
            [
              {
                "flex": 1,
              },
              undefined,
            ]
          }
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "alignSelf": "center",
                  "height": "100%",
                  "justifyContent": "center",
                  "minWidth": 25,
                  "position": "absolute",
                  "width": "100%",
                },
                {
                  "opacity": 0,
                },
              ]
            }
          >
            <Text>
              Icon
            </Text>
          </View>
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "alignSelf": "center",
                  "height": "100%",
                  "justifyContent": "center",
                  "minWidth": 25,
                  "position": "absolute",
                  "width": "100%",
                },
                {
                  "opacity": 1,
                },
              ]
            }
          >
            <Text>
              Icon
            </Text>
          </View>
        </View>
        <Text
          numberOfLines={1}
          style={
            [
              {
                "backgroundColor": "transparent",
                "textAlign": "center",
              },
              {
                "color": "#BBDEFB",
              },
              {
                "fontSize": 10,
              },
              {
                "fontSize": 10,
                "paddingBottom": 5,
              },
            ]
          }
        >
          Me
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
